File: Sliding_Window/PythonSolutions/LongestSubstringWithoutRepeatingCharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        maxLength =0
        charSet = set()
        L =0

        for R in range(n):
            if s[R] not in charSet:
                charSet.add(s[R])
                maxLength = max(maxLength, R -L +1)
            else:
                while s[R] in charSet:
                    charSet.remove(s[L])
                    L +=1
                charSet.add(s[R])
        return maxLength

File: Sliding_Window/JavaSolutions/LongestSubstringWithoutRepeatingCharacters.java
package Sliding_Window.JavaSolutions;

import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxLength = 0;
        HashSet<Character> charSet = new HashSet<>();
        int L = 0;

        for (int R = 0; R < n; R++) {
            char currentChar = s.charAt(R);

            // If current character is already in the set, shrink the window
            while (charSet.contains(currentChar)) {
                charSet.remove(s.charAt(L));
                L++;
            }

            charSet.add(currentChar);
            maxLength = Math.max(maxLength, R - L + 1);
        }

        return maxLength;
    }
}
